src/pyodide/internal/workers.py

```diff
@@ -141,62 +158,248 @@ def json(
+        try:
+            yield buf.data
+            return
+        finally:
+            buf.release()
+            px.destroy()
+    raise TypeError(f"Don't know how to handle {type(e)} for Blob()")
+
+
+def _is_iterable(obj):
+    try:
+        iter(obj)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+BlobValue = (
+    "str | bytes | js.ArrayBuffer | js.TypedArray | js.DataView | js.Blob | Blob"
+)
+
+
+class BlobEnding(StrEnum):
+    TRANSPARENT = "transparent"
+    NATIVE = "native"
+
+
+class Blob:
+    def __init__(
+        self,
+        blob_parts: "Iterable[BlobValue] | BlobValue",
+        content_type: str | None = None,
+        endings: BlobEnding | str | None = None,
+        filename: str | None = None,
+        last_modified: int | None = None,
+    ):
+        if endings:
+            endings = str(endings)
+
+        is_single_item = not _is_iterable(blob_parts)
+        if is_single_item:
+            # Inherit the content_type, filename and lastModified if we have a
+            # single item.
+            if not content_type and isinstance(blob_parts, Blob):
+                content_type = blob_parts.content_type
+            if hasattr(blob_parts, "constructor") and (
+                blob_parts.constructor.name in ("Blob", "File")
+            ):
+                if not content_type:
+                    content_type = blob_parts.type
+                if blob_parts.constructor.name == "File":
+                    if not filename:
+                        filename = blob_parts.name
+                    if not last_modified:
+                        last_modified = blob_parts.lastModified
+
+            # Otherwise create a new Blob below.
+            blob_parts = [blob_parts]
+
+        with ExitStack() as stack:
+            args = [stack.enter_context(_make_blob_entry(e)) for e in blob_parts]
+            with _manage_pyproxies() as pyproxies:
+                if filename or last_modified:
+                    filename = "" if filename is None else filename
+                    self._js_blob = js.File.new(
+                        to_js(args, pyproxies=pyproxies),
+                        filename,
+                        type=content_type,
+                        endings=endings,
+                        lastModified=last_modified,
+                    )
+                else:
+                    self._js_blob = js.Blob.new(
+                        to_js(args, pyproxies=pyproxies),
+                        type=content_type,
+                        endings=endings,
+                    )
+
+    @property
+    def size(self) -> int:
+        return self._js_blob.size
+
+    @property
+    def content_type(self) -> str:
+        return self._js_blob.type
+
+    @property
+    def name(self) -> str:
+        if self._js_blob.constructor.name != "File":
+            raise TypeError(
+                "Underlying Blob type is not a File and so has no file name."
+            )
```

Comment:
It's not obvious to me that it's a good idea to unify `Blob` and `File`, but if we do, it's not obvious to me we should raise `TypeError` here. Another option would be to return `None` here. In fact, if you just did:
```py
return self._js_blob.name
```
then it would return `None` for `Blob`s.